### PR TITLE
185791342 ce assessments services

### PR DESCRIPTION
**Pull request by ZachCMP**
https://www.pivotaltracker.com/story/show/185821067

Relies on frontend PR: https://github.com/greenriver/hmis-frontend/pull/412

### CODE REVIEW

File: drivers/hmis/app/models/hmis/form/form_processor.rb [hunk 1]
```diff
@@ -30,6 +30,8 @@ class Hmis::Form::FormProcessor < ::GrdaWarehouseBase
   belongs_to :youth_education_status, class_name: 'Hmis::Hud::YouthEducationStatus', optional: true, autosave: true
   belongs_to :employment_education, class_name: 'Hmis::Hud::EmploymentEducation', optional: true, autosave: true
   belongs_to :current_living_situation, class_name: 'Hmis::Hud::CurrentLivingSituation', optional: true, autosave: true
+  belongs_to :ce_assessment, class_name: 'Hmis::Hud::Assessment', optional: true, autosave: true
+  belongs_to :ce_event, class_name: 'Hmis::Hud::Event', optional: true, autosave: true
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/form/form_processor.rb]: This relationship doesn't exist in the database, so we should remove this.
We would only need this if `Hmis::Hud::CustomAssessment` are generating Events/Assessments, which they're not (yet). 

For "record-editing forms" that use the SubmitForm mutation, the FormProcessor record isn't actually persisted.

File: drivers/hmis/app/models/hmis/hud/event.rb [hunk 2]
```diff
@@ -15,6 +15,7 @@ class Hmis::Hud::Event < Hmis::Hud::Base
   belongs_to :client, **hmis_relation(:PersonalID, 'Client')
   belongs_to :user, **hmis_relation(:UserID, 'User'), inverse_of: :events
   belongs_to :data_source, class_name: 'GrdaWarehouse::DataSource'
+  has_one :project, through: :enrollment
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/hud/event.rb]: Why is this added, and is it used anywhere here?
I'm wondering because this relation will return `nil` for events rendered at WIP enrollments. So, we should add a comment about that at least, and be mindful when using.

**Comment by ZachCMP** [comment]
@gigxz Addressed your feedback

File: drivers/hmis/app/models/hmis/hud/event.rb (hunk 2 above)
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/hud/event.rb]: This was added so that permissions can resolve through the project, but good point on the WIP issue. I just added another case in the loader to use the enrollment directly instead of doing it this way.

File: drivers/hmis/app/models/hmis/entity_access_loader_factory.rb [hunk 3]
```diff
@@ -79,6 +79,8 @@ def resolve_association(entity, safety:, &block)
       else
         block.call(entity, :project)
       end
+    when Hmis::Hud::Event
+      block.call(entity, :enrollment)
     else
       resolve_through_project(entity, &block)
     end
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/entity_access_loader_factory.rb]: OK, I see. This type of case would also be needed for Hmis::Hud::Assessment as well.

I think instead of this, the correct approach that Theron introduced is to `include ::Hmis::Hud::Concerns::ClientProjectEnrollmentRelated` in the `Hmis::Hud::Assessment` and `Hmis::Hud::Event` model. That concern adds a relationship to project that goes through a view which includes wip enrollments, so all projects should be resolvable.

Comment (gigxz) [reply drivers/hmis/app/models/hmis/entity_access_loader_factory.rb]: I can make this change and merge since you'll be offline tm

File: drivers/hmis/lib/form_data/default/records/ce_assessment.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          "mapping": {
+            "field_name": "assessmentDate"
+          }
+        },
+        {
+          "type": "TEXT",
+          "required": true,
+          "link_id": "4.19.2",
+          "text": "Assessment Location",
+          "mapping": { "field_name": "assessmentLocation" }
```
Comment (gigxz) [inline drivers/hmis/lib/form_data/default/records/ce_assessment.json:21]: >  It is recommended that a system administrator managed list is used for this
field. If such functionality doesn’t exist in the HMIS, a text box must be provided.

@eanders do you know what this location is supposed to be, is it the CoC? or some other custom managed list of locations?

Comment (eanders) [reply drivers/hmis/lib/form_data/default/records/ce_assessment.json:21]: I don't know.  I believe we've seen everything from the CoC Code to the project or organization name to a specific address.  My gut would be project names...

File: drivers/hmis/lib/form_data/default/records/ce_event.json [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+              "answer_code": "REFERRAL_TO_POST_PLACEMENT_FOLLOW_UP_CASE_MANAGEMENT"
+            }
+          ]
+        },
+        {
+          "type": "TEXT",
+          "required": false,
+          "link_id": "4.20.C",
+          "text": "Location of Crisis Housing or Permanent Housing Referral",
+          "mapping": { "field_name": "locationCrisisOrPhHousing" },
```
Comment (gigxz) [inline drivers/hmis/lib/form_data/default/records/ce_event.json:59]: (No action needed, leaving this here for reference..)
Screen Shot 2023-08-16 at 8 06 15 PM

Comment (gigxz) [reply drivers/hmis/lib/form_data/default/records/ce_event.json:59]: @awisema , this looks like one place where the "receives CE referrals" flag is used.

File: drivers/hmis/lib/form_data/default/records/ce_assessment.json (hunk 4 above)
Comment (gigxz) [reply drivers/hmis/lib/form_data/default/records/ce_assessment.json:21]: Huh okay. None of those really make sense to me, because you should be able to auto determine that based on the Enrollment where you're recording this. Unless the location is typically at a _different_ project? I think we leave as text for now, and ask.

**Review by gigxz — APPROVED** [review]

File: drivers/hmis/lib/form_data/default/records/ce_assessment.json (hunk 4 above)
Comment (ZachCMP) [reply drivers/hmis/lib/form_data/default/records/ce_assessment.json:21]: Oh, yeah, sorry, I should have mentioned that. That field is supposed to be a pick list of projects but the logic seemed thorny so I was going to circle back around on it and ping you about how we should do it after I was done with everything else.
